loadgen/loadgen.cc

```diff
@@ -1076,171 +1073,75 @@ struct LogOutputs {
-    LoadSamplesToRam(qsl, performance_set.set);
-
-    PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
-        sut, u_bound, performance_set, sequence_gen));
-    PerformanceSummary perf_summary{sut->Name(), u_bound, std::move(pr)};
-
-    qsl->UnloadSamplesFromRam(performance_set.set);
-
-    if (!perf_summary.PerfConstraintsMet(&msg)) {
-      LogDetail([msg](AsyncDetail& detail) {
-        detail(
-            "FoundBinaries: Initial lower bound does not satisfy performance "
-            "constraints, msg: " +
-            msg);
-      });
-      return std::make_pair(l_bound, l_bound);
-    }
-  }
-
-  // Find upper bound settings.
-  while (true) {
-    // Generate loadable sets based on upper bound.
-    // Re-generating is required to cover the MultiStream cases (padded
-    // samples).
-    std::vector<loadgen::LoadableSampleSet> loadable_sets(
-        loadgen::GenerateLoadableSets(qsl, u_bound));
-
-    const LoadableSampleSet& performance_set = loadable_sets.front();
-    LoadSamplesToRam(qsl, performance_set.set);
-
-    PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
-        sut, u_bound, performance_set, sequence_gen));
-    PerformanceSummary perf_summary{sut->Name(), u_bound, std::move(pr)};
-
-    qsl->UnloadSamplesFromRam(performance_set.set);
+  // Generate loadable sets based on upper bound.
+  // Re-generating is required to cover the MultiStream cases (padded
+  // samples).
+  std::vector<loadgen::LoadableSampleSet> loadable_sets(
+      loadgen::GenerateLoadableSets(qsl, u_bound_settings));
+  const LoadableSampleSet& performance_set = loadable_sets.front();
+  LoadSamplesToRam(qsl, performance_set.set);
 
-    if (!perf_summary.PerfConstraintsMet(&msg)) {
-      break;
-    }
+  PerformanceResult pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+      sut, u_bound_settings, performance_set, sequence_gen));
+  PerformanceSummary perf_summary{sut->Name(), u_bound_settings, std::move(pr)};
 
-    find_peak_performance::SetPerformanceField<scenario>(l_bound, u_bound);
-    find_peak_performance::WidenPerformanceField<scenario>(u_bound);
+  qsl->UnloadSamplesFromRam(performance_set.set);
 
-    LogDetail(
-        [l_field =
-             find_peak_performance::ToStringPerformanceField<scenario>(l_bound),
-         r_field = find_peak_performance::ToStringPerformanceField<scenario>(
-             u_bound)](AsyncDetail& detail) {
-          detail("FindBoundaries: Widen fields to [" + l_field + ", " +
-                 r_field + ")");
-        });
+  std::string tmp;
+  if (!perf_summary.PerfConstraintsMet(&tmp)) {
+    return std::make_pair(base_perf_summary, perf_summary);
+  } else {
+    return FindBoundaries<scenario>(sut, qsl, sequence_gen, perf_summary);
   }
-
-  return std::make_pair(l_bound, u_bound);
 }
 
 template <TestScenario scenario>
-TestSettingsInternal BinarySearchToFindPeakPerformance(
+PerformanceSummary FindPeakPerformanceBinarySearch(
```

Comment:
Change the return type to `PerformanceSummary` to reuse the result of execution. (The previous version re-run `IssueQueries` to get `PerformanceSummary` using the result value `TestSettingsInternal`)